plugins/Transitions/API.php

```diff
@@ -60,7 +60,7 @@ public function getTransitionsForPageUrl($pageUrl, $idSite, $period, $date, $seg
      * @return array
      * @throws Exception
      */
-    public function getTransitionsForAction($actionName, $actionType, $idSite, $period, $date,
+    public function getTransitionsForAction(string $actionName, string $actionType, $idSite, $period, $date,
```

Comment:
That would be nice, but some parameters like `$idSite` can legitimately have string and numeric values, eg. `1` or `all`, so it can be problematic to apply type hints without some degree of certainty that the parameter should only accept a single  type. Given that this was a time-boxed quick fix outside of the current sprint I'm reluctant to start wielding the can opener on this particular issue. :worm::worm::worm::worm:

Since many API methods receive the same parameters (idSite, period, date) it would be good to standardize on the strict types (if any) we can use for these across the whole API, and work towards having strict type on all API methods, but this will need some planning and is probably a 6.x thing. :slightly_smiling_face:

Reply:
@michalkleiner We can add such typehints to our API methods at any point, unless it would cause some sort of breaking change. Typecasts in API methods were not possible before Matomo 5, as our API processor wasn't able to handle them. Before parameters were actually always provided as `string`, `array` or `null`. Currently the most API methods don't use type hints, as it's hard to say what might break when simply adding them anywhere. My approach would be to add them whenever we need to work on a certain API method or add a new one and the types can be clearly defined.
This will also help use to fix certain bug reports that come in over and over again, where a user provides an invalid value type and the API throws a PHP error, as the type is handled correctly.